topologies/all/eos-reset.py

```diff
@@ -0,0 +1,342 @@
+        tmp_veos_ips.append(re_veos[cur_veos]['internal_ip'])
+        # Push bare config to remote startup-config and reboot device
+        if pushBareConfig(cur_veos, re_veos[cur_veos]['ip'], tmp_veos_config):
+            DEVREBOOT = True
+            dev_reboot_status[cur_veos] = {"status":False, "IP":re_veos[cur_veos]['internal_ip']}
+    pS("OK", "Configs pushed to: {0}".format(", ".join(vdevs)))
+    if DEVREBOOT:
+        pS("INFO", "Devices rebooting, waiting for devices to come back online")
+    # Create connection to CVP
+    cvp_clnt = conCVP()
+    # Restore CVP Configlets back to repo base
+    restoreConfiglets(cvp_clnt, configlet_location)
+    restoreContainers(cvp_clnt)
+    cvp_clnt.saveTopology()
+    # Check for any pending tasks from configlet restore
+    cvp_clnt.getAllTasks("pending")
+    if cvp_clnt.tasks['pending']:
+        pS("OK", "Executing all tasks")
+        cvp_clnt.execAllTasks("pending")
+        pS("OK", "Tasks executed")
+    # Check for failed tasks and cancel them.
+    cvp_clnt.getAllTasks("failed")
+    if cvp_clnt.tasks['failed']:
+        pS("OK", "Cancelling previously failed tasks")
+        cvp_clnt.cancelTasks("failed")
+    # Check to verify all devices to be reset have booted up
+    for vdevR in dev_reboot_status:
+        while not dev_reboot_status[vdevR]['status']:
+            ip_res = cvp_clnt.ipConnectivityTest(dev_reboot_status[vdevR]['IP'])
+            if 'data' in ip_res:
+                pS("OK", "{0} is back online".format(vdevR))
+                dev_reboot_status[vdevR]['status'] = True
+            else:
+                pS("INFO", "Waiting for {0} to comeback online... Checking again in {1} seconds.".format(vdevR, DELAYTIMER))
+                sleep(DELAYTIMER)
+    # Grab container mappings
+    eos_cnt_map = eosContainerMapper(cvp_yaml['cvp_info']['containers'])
+    # Get EOS info for devices needing to be updated
+    eos_info = getEosDevice(atd_yaml['topology'], atd_yaml['nodes']['veos'], eos_cnt_map, vdevs)
+    # Add all devices to be reset into CVP
+    cvp_clnt.addDeviceInventory(tmp_veos_ips)
+    # Iterate through EOS Objects that need to be reset
+    for eos in eos_info:
+        provisionDevice(cvp_clnt, eos)
+    cvp_clnt.saveTopology()
+    cvp_clnt.getAllTasks("pending")
+    pS("OK", "Executing all tasks")
+    cvp_clnt.execAllTasks("pending")
+    pS("OK", "Tasks executed")
+    sleep(5)
+    # All tasks should complete successfully, but need to check for failed tasks
+    f_vdevs = [] # List to contain failed devices
+    cvp_clnt.getAllTasks("failed")
+    pS("OK", "Checking for failed tasks")
+    if cvp_clnt.tasks['failed']:
+        pS("INFO", "Found failed tasks")
+        tmp_veos_ips = []
+        # Iterate through list of failed tasks, remove device and add hostname to f_vdevs for re-provisioning
+        for ftask in cvp_clnt.tasks['failed']:
```

Comment:
This is why, I had to put a check in to see if there were any failed tasks, since depending on the situation a task could fail.  So I wanted to capture it and re-try that device.  So it does a search for any failed tasks.  Make sure any pending and failed tasks have been cancelled first, then try again.  I can add in a check to make sure it is only re-doing a failed task for the specified device.